File: Next_Right_Pointers.java
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public Node connect(Node root) {
        helper(root);
        return root;
    }
    
    public void helper(Node root){
        if(root==null) return;
        
        if(root.left!=null){
            root.left.next=root.right;
        }
        
        if(root.next!=null && root.right!=null){
            root.right.next=root.next.left;
        }
        
        helper(root.left);
        helper(root.right);
    }
}

File: nextRightPointers.py
#Time Complexity: O(n)
#Space Complexity: O(n)
class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        q = deque()
        q.append(root)
        
        if root is None:
            return root
        
        while q:
            size = len(q)
            for i in range(size):
                el = q.popleft()
                if i < size-1:
                    el.next = q[0]
                    
                if el.left:
                    q.append(el.left)
                    
                if el.right:
                    q.append(el.right)
                    
        return root
                
                